src/librustc_parse/parser/item.rs

```diff
@@ -858,16 +853,19 @@ impl<'a> Parser<'a> {
                     self.error_on_foreign_const(span, ident);
                     ForeignItemKind::Static(a, Mutability::Not, b)
                 }
-                _ => {
-                    let span = self.sess.source_map().def_span(span);
-                    self.struct_span_err(span, "item kind not supported in `extern` block").emit();
-                    return None;
-                }
+                _ => return self.error_bad_item_kind(span, &kind, "`extern` block"),
             };
             Some(P(Item { attrs, id, span, vis, ident, defaultness, kind, tokens }))
         }))
     }
 
+    fn error_bad_item_kind<T>(&self, span: Span, kind: &ItemKind, ctx: &str) -> Option<T> {
+        let span = self.sess.source_map().def_span(span);
+        let msg = format!("{} not supported in {}", kind.descr(), ctx);
```

Comment:
I considered this, but it doesn't work out with e.g. `type alias` because you get "type aliass"; I would prefer not to roll a second version of `fn descr`. I could do some sorta hack that checks if the last character ends with "s" perhaps... but that feels ugly wrt. prospective internationalization?

Reply:
```suggestion
        let msg = format!("{} is not supported in {}s", kind.descr(), ctx);
```
then.

Reply:
👍